cookbooks/bcpc-hadoop/libraries/utils.rb

```diff
@@ -186,76 +186,66 @@ def znode_exists?(znode_path, zk_host = 'localhost:2181')
-      |hst| { 'hostname' => hst[:fqdn]} }
-
-  node.default[:bcpc][:hadoop][:hb_hosts] = 
-    cd.fetch_cluster_def.select { |hst| hst[:runlist].include? srvc2role['hbase_head'] }.map {
-      |hst| { 'hostname' => hst[:fqdn]} }
-
-  # different flavors of hive
-  node.default[:bcpc][:hadoop][:hive_hosts] = 
-      cd.fetch_cluster_def.select { |hst| hst[:runlist].include? srvc2role['hive_server'] }.map {
-      |hst| { 'hostname' => hst[:fqdn]} }
-
-  # BCPC-Hadoop-Head-MapReduce
-  node.default[:bcpc][:hadoop][:oozie_hosts]  = 
-    cd.fetch_cluster_def.select { |hst| hst[:runlist].include? srvc2role['oozie_server'] }.map {
-      |hst| { 'hostname' => hst[:fqdn]} }
-  
-  # every datanode
-  node.default[:bcpc][:hadoop][:httpfs_hosts] = 
-    cd.fetch_cluster_def.select { |hst| hst[:runlist].include? srvc2role['hadoop_worker']  }.map {
-      |hst| { 'hostname' => hst[:fqdn]} }
-
-  # Worker
-  node.default[:bcpc][:hadoop][:rs_hosts] = 
-    cd.fetch_cluster_def.select { |hst| hst[:runlist].include? srvc2role['hadoop_worker']  }.map {
-      |hst| { 'hostname' => hst[:fqdn]} }
-
-  # BCPC-Hadoop-Head
-  node.default[:bcpc][:hadoop][:mysql_hosts] = 
-    cd.fetch_cluster_def.select { |hst| hst[:runlist].include? srvc2role['hadoop_head'] }.map {
-      |hst| { 'hostname' => hst[:fqdn]} }
+  hosts = node.run_state['cluster_def'].fetch_cluster_def
+
+  # host search helper lambdas
+  runs_role = Proc.new { |host, role| role && host[:runlist].include?(role) }
+  runs_recipe = Proc.new { |host, recipe| recipe && host[:runlist].include?(recipe) }
+
+  # mapped host objects
+  to_host = Proc.new do |host| {
+    'hostname' => host[:fqdn],
+    'node_number' => host[:node_id],
+    'zookeeper_myid' => nil
+  } end
+
+  node['bcpc']['hadoop']['services'].each do |name, service|
+    *keys, last = ['bcpc', 'hadoop', *service['key']]
+    keys.inject(node.default, :fetch)[last] =
+      hosts.select do |h|
+        runs_role.call(h, service['role']) ||
+        runs_recipe.call(h, service['recipe'])
+      end.map do |h|
+        to_host.call(h)
+      end
+  end
+
+  # set the oozie_url
+  oozie_hosts = node['bcpc']['hadoop']['oozie_hosts']
+  vip_host = float_host(node['bcpc']['management']['viphost'])
+  first_host = float_host(oozie_hosts.first['hostname'])
+  oozie_ha_port = node['bcpc']['ha_oozie']['port']
+  oozie_port = node['bcpc']['hadoop']['oozie_port']
+
+  node.default['bcpc']['hadoop']['oozie_url'] =
+    if oozie_hosts.length > 1
+      # high-availability
+      "http://#{vip_host}:#{oozie_ha_port}/oozie"
+    elsif oozie_hosts.length == 1
+      # single oozie host
+      "http://#{first_host}/#{oozie_port}/oozie"
+    end
+
+  # set the resourcemanager_url (rm_address)
+  rm_hosts = node['bcpc']['hadoop']['rm_hosts']
+  first_host = float_host(rm_hosts.first['hostname'])
+  rm_port = node['bcpc']['hadoop']['yarn']['resourcemanager']['port']
+
+  node.default['bcpc']['hadoop']['rm_address'] =
+    if rm_hosts.length > 1
```

Comment:
Add comment:

```
# FIXME if we just use plain HA ResourceManagers in yarn_site.rb we can simply remove this if statement
```

https://github.com/bloomberg/chef-bach/blob/master/cookbooks/bcpc-hadoop/recipes/yarn_site.rb#L37

Reply:
No we should not always assume the RM's are HA; test-kitchen style tests will not be.

Reply:
I don't understand why not.  The failover proxy provider clients can support having only one RM defined behind its configuration.  test-kitchen should still work.

Reply:
Ah I had not realized one could run a single RM without triggering an error that multiple were needed. It does still pull in Zookeeper though.